// contentful-migration/built/lib/intent/tag-create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("./base-intent"));
const tag_create_1 = require("../action/tag-create");
const chalk_1 = __importDefault(require("chalk"));
class TagCreateIntent extends base_intent_1.default {
    isTagIntent() {
        return true;
    }
    getTagId() {
        return this.payload.tagId;
    }
    getTagVisibility() {
        return this.payload.tagVisibility;
    }
    isTagCreate() {
        return true;
    }
    groupsWith() {
        return false;
    }
    endsGroup() {
        return false;
    }
    shouldPublish() {
        return false;
    }
    requiresAllTags() {
        return true;
    }
    toActions() {
        return [
            new tag_create_1.TagCreateAction(this.getTagId(), this.getTagVisibility())
        ];
    }
    toPlanMessage() {
        return {
            heading: chalk_1.default `Create Tag {bold.yellow ${this.getTagId()}} with visibility: "${this.getTagVisibility()}"`,
            details: [],
            sections: []
        };
    }
}
exports.default = TagCreateIntent;
