// contentful-migration/built/lib/intent/tag-delete.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("./base-intent"));
const tag_delete_1 = require("../action/tag-delete");
const chalk_1 = __importDefault(require("chalk"));
class TagDeleteIntent extends base_intent_1.default {
    isTagIntent() {
        return true;
    }
    getTagId() {
        return this.payload.tagId;
    }
    isTagDelete() {
        return true;
    }
    groupsWith() {
        return false;
    }
    endsGroup() {
        return true;
    }
    shouldSave() {
        return false;
    }
    shouldPublish() {
        // TODO: Do we need this in general in tag classes?
        return false;
    }
    requiresAllTags() {
        return true;
    }
    toActions() {
        return [
            new tag_delete_1.TagDeleteAction(this.getTagId())
        ];
    }
    toPlanMessage() {
        return {
            heading: chalk_1.default `Delete Tag {bold.yellow ${this.getTagId()}}`,
            details: [],
            sections: []
        };
    }
}
exports.default = TagDeleteIntent;
