// contentful-migration/built/lib/intent-validator/entry-set-tags.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_validator_1 = __importDefault(require("./schema-validator"));
const Joi = __importStar(require("joi"));
class EntrySetTagsIntentValidator extends schema_validator_1.default {
    constructor() {
        super(...arguments);
        this.article = 'a';
        this.displayName = 'tags update for entry';
    }
    appliesTo(step) {
        return step.isEntrySetTags();
    }
    // TODO The propertyNameToValidate method is supposed to be
    // refactored, see other intents.
    get propertyNameToValidate() {
        return 'entryTransformationForTags';
    }
    get schema() {
        return {
            contentType: Joi.string().required(),
            from: Joi.array().items(Joi.string()).required(),
            setTagsForEntry: Joi.func().required()
        };
    }
}
exports.default = EntrySetTagsIntentValidator;
