// contentful-migration/built/lib/intent-validator/field-update.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_validator_1 = __importDefault(require("./schema-validator"));
const Joi = __importStar(require("joi"));
class FieldUpdateStepValidator extends schema_validator_1.default {
    constructor() {
        super(...arguments);
        this.article = 'a';
        this.displayName = 'field';
    }
    appliesTo(step) {
        return step.isFieldUpdate();
    }
    get schema() {
        return {
            id: Joi.string().required(),
            newId: Joi.string().required(),
            name: Joi.string().required(),
            type: Joi.string().required(),
            localized: Joi.boolean().required().strict(),
            required: Joi.boolean().required().strict(),
            validations: Joi.array().required(),
            defaultValue: Joi.object().required().strict(),
            disabled: Joi.boolean().required().strict(),
            omitted: Joi.boolean().required().strict(),
            deleted: Joi.boolean().required().strict(),
            linkType: Joi.string().required(),
            items: Joi.object().required()
        };
    }
}
exports.default = FieldUpdateStepValidator;
